src/nvim/api/vim.c

```diff
@@ -164,6 +164,83 @@ Object vim_eval(String str, Error *err)
   return rv;
 }
 
+/// Call the given function with the given arguments stored in an array.
+///
+/// @param functionName The function to call
+/// @param args the functions arguments packed in an Array
+/// @param self optional self dictionary that may be used by user functions to
+///        emulate object-oriented programming
+/// @param[out] err Details of an error that may have occurred
+/// @return The result of the function call if it was successful
+Object vim_call_function(String functionName, Array args, Object self,
+                         Error *err)
+  FUNC_ATTR_DEFERRED
+{
+  Object rv = OBJECT_INIT;
+  if (args.size > MAX_FUNC_ARGS) {
+    api_set_error(err, Validation,
+      _("Function called with too many arguments."));
+    return rv;
+  }
+
+  typval_T optional_self_dict;
+  dict_T *self_dict = NULL;
+  if (self.type == kObjectTypeNil) {
+    self_dict = NULL;
+  } else if (self.type == kObjectTypeDictionary
+             || self.type == kObjectTypeArray) {
+    // XXX The tests somehow cause the object to be an array
+    self.type = kObjectTypeDictionary;
```

Comment:
This is completely incorrect. Due to the reasons explained in my message tests do pass an array in place of a dictionary. The only reason this code works is that it happens when array is empty in tests. You will get a crash if you happen to pass real array here and change its type this way.